File: code/task3predict.py
from collections import defaultdict
from itertools import combinations
from pyspark import SparkContext
from math import sqrt
import json
import random
import sys
import time


class T3pred:
    def __init__(self) -> None:
        self.ipf = sys.argv[1]
        self.testfile = sys.argv[2]
        self.modelfile = sys.argv[3]
        self.outfile = sys.argv[4]
        self.cf_type = sys.argv[5]

    @staticmethod
    def get_usersimmap(rows):
        usm = {}

        for row in rows:
            user1, user2 = list(sorted((row["u1"], row["u2"])))
            sim = row["sim"]
            usm[(user1, user2)] = sim
        return usm

    @staticmethod
    def get_all_userids(usm):
        user_ids = set()
        for uid1, uid2 in usm:
            user_ids.add(uid1)
            user_ids.add(uid2)
        return user_ids

    @staticmethod
    def genubr(ubrlist):
        ubr = defaultdict(dict)
        for k, v in ubrlist.items():
            ubr[k][v[0]] = v[1]
        return ubr

    @staticmethod
    def get_item_prediction(uid, bid, ratings_list, modelMap):
        NEIGHBORS = 5
        valid_neighbors = []
        
        if uid not in ratings_list:
            return 0
        for bid2, rating in ratings_list[uid]:
            if tuple(sorted([bid, bid2])) in modelMap:
                valid_neighbors.append((rating, modelMap[tuple(sorted([bid, bid2]))]))
        
        valid_neighbors.sort(key=lambda x: x[1], reverse=True)
        valid_neighbors = valid_neighbors[:NEIGHBORS]

        top = bot = 0
        for x in valid_neighbors:
            top += x[0] * x[1]
            bot += abs(x[1])
        return 0 if (top == 0 or bot == 0) else top / bot

    @staticmethod
    def get_user_prediction(uid, bid, ratings_list, modelMap, ur_avg, ubMap):
        NEIGHBORS = 5
        valid_neighbors = []
        
        if bid not in ratings_list:
            return 0
        for uid2, rating in ratings_list[bid]:
            key = tuple(sorted([uid, uid2]))
            if key in modelMap:
                normal_rating = rating - ur_avg[uid2]
                valid_neighbors.append((normal_rating, modelMap[key]))
        
        valid_neighbors.sort(key=lambda x: x[1], reverse=True)
        valid_neighbors = valid_neighbors[:NEIGHBORS]

        top = bot = 0
        for x in valid_neighbors:
            top += x[0] * x[1]
            bot += abs(x[1])
        return 0 if (top == 0 or bot == 0) else (ur_avg[uid] + top / bot)

    def run_item_based(self):
        sc = SparkContext.getOrCreate()
        sc.setLogLevel("OFF")

        # {(b1, b2): sim}
        modelMap = sc.textFile(self.modelfile).map(lambda row: json.loads(row)).map(lambda row: (tuple(sorted([row["b1"], row["b2"]])), row["sim"])).collectAsMap()
        # {u1: {b1: 5, b2: 3}}
        textRDD = sc.textFile(self.ipf).map(lambda row: json.loads(row))
        ubr = textRDD.map(lambda row: (row["user_id"], (row["business_id"], row["stars"]))).groupByKey().map(lambda row: (row[0], set(row[1]))).collectAsMap()

        testRDD = sc.textFile(self.testfile).map(lambda row: json.loads(row))
        test_pairs = testRDD.map(lambda row: (row["user_id"], row["business_id"])).collect()
        
        with open(self.outfile, "w+") as f:
            for pair in test_pairs:
                uid = pair[0]
                bid = pair[1]
                rating = T3pred.get_item_prediction(uid, bid, ubr, modelMap)
                if rating:
                    f.write(json.dumps({"user_id": uid, "business_id": bid, "stars": rating}) + "\n")

    def run_user_based(self):
        sc = SparkContext.getOrCreate()
        sc.setLogLevel("OFF")

        # {(u1, u2): sim}
        modelMap = sc.textFile(self.modelfile).map(lambda row: json.loads(row)).map(lambda row: (tuple(sorted([row["u1"], row["u2"]])), row["sim"])).collectAsMap()
        
        # {u1: {b1: 5, b2: 3}}
        textRDD = sc.textFile(self.ipf).map(lambda row: json.loads(row))
        textRDD.cache()

        bur = textRDD.map(lambda row: (row["business_id"], (row["user_id"], row["stars"]))).groupByKey().map(lambda row: (row[0], set(row[1]))).collectAsMap()
        
        ur_avg = textRDD.map(lambda row: (row["user_id"], row["stars"])).groupByKey().mapValues(lambda row: list(row)).mapValues(lambda row: sum(row) / len(row)).collectAsMap()
        ubMap = textRDD.map(lambda row: ((row["user_id"], row["business_id"]), row["stars"])).collectAsMap()
        
        testRDD = sc.textFile(self.testfile).map(lambda row: json.loads(row))
        test_pairs = testRDD.map(lambda row: (row["user_id"], row["business_id"])).collect()
        
        with open(self.outfile, "w+") as f:
            for uid, bid in test_pairs:
                rating = T3pred.get_user_prediction(uid, bid, bur, modelMap, ur_avg, ubMap)
                if rating:
                    f.write(json.dumps({"user_id": uid, "business_id": bid, "stars": rating}) + "\n")

    def run(self):
        if self.cf_type == "user_based":
            self.run_user_based()
        else:
            self.run_item_based()


if __name__ == "__main__":
    t3 = T3pred()

    st = time.time()
    t3.run()
    et = time.time()
    print("Total time taken: {}".format(et - st))

File: hw3/src/main/scala/task3predict.scala
import org.apache.spark.sql.SparkSession
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.rdd.RDD
import org.apache.spark.storage.StorageLevel
import org.json4s._
import org.json4s.jackson.JsonMethods._

import scala.math.log10
import scala.util.control.Breaks._
import scala.util.Random._
import collection.mutable.{ArrayBuffer, HashMap, HashSet}
import java.io.{File, PrintWriter}
import scala.io.Source
import scala.math.Ordering.Implicits._
import scala.math.sqrt
import scala.util.Random


object task3predict {
  implicit val formats = DefaultFormats

  def sortIDs(id1: String, id2: String): (String, String) = {
    if (id2 < id1) {
      return (id2, id1)
    }
    return (id1, id2)
  }

  def loadItemRow(row: JValue): ((String, String), Double) = {
    var b1 = (row\"b1").extract[String]
    var b2 = (row\"b2").extract[String]
    val sim = (row\"sim").extract[Double]

    if (b2 < b1) {
      var temp = b1
      b1 = b2
      b2 = temp
    }

    return ((b1, b2), sim)
  }

  def ubrRow(row: JValue): (String, (String, Double)) = {
    var u1 = (row\"user_id").extract[String]
    var b1 = (row\"business_id").extract[String]
    val stars = (row\"stars").extract[Double]

    return (u1, (b1, stars))
  }

  def getItemPred(uid: String, bid: String, ubr: scala.collection.Map[String, Set[(String, Double)]], modelMap: scala.collection.Map[(String, String), Double]): Double = {
    val NEIGHBORS = 10

    var validNeighbors = new ArrayBuffer[(Double, Double)]

    if (!ubr.contains(uid)) {
      return 0.0
    }
    for (pair <- ubr(uid)) {
      val bid2 = pair._1
      val rating = pair._2
      val bids = sortIDs(bid2, bid)
      if (modelMap.contains(bids)) {
        validNeighbors += ((rating, modelMap(bids)))
      }
    }

    validNeighbors.sortBy(x => -x._2)
    validNeighbors = validNeighbors.take(NEIGHBORS)

    var top = 0.0
    var bot = 0.0

    for (x <- validNeighbors) {
      top += (x._1) * (x._2)
      bot += math.abs(x._2)
    }

    if (top == 0.0 || bot == 0.0) {
      return 0
    }
    return (top.toDouble / bot.toDouble)
  }

  def itemBased(args: Array[String]) = {
    val trainfile = args(0)
    val testfile = args(1)
    val modelfile = args(2)
    val outfile = args(3)

    val sc = SparkContext.getOrCreate()
    sc.setLogLevel("OFF")

    val modelMap = sc.textFile(modelfile)
      .map(row => parse(row))
      .map(row => loadItemRow(row))
      .collectAsMap()

    val ubr = sc.textFile(trainfile)
      .map(row => parse(row))
      .map(row => ubrRow(row))
      .groupByKey()
      .mapValues(value => value.toSet)
      .collectAsMap()

    val testPairs = sc.textFile(testfile)
      .map(row => parse(row))
      .map(row => ((row\"user_id").extract[String], (row\"business_id").extract[String]))
      .collect()

    val out = new PrintWriter(new File(outfile))

    for (pair <- testPairs) {
      val uid = pair._1
      val bid = pair._2
      val rating = getItemPred(uid, bid, ubr, modelMap)
      if (rating != 0) {
        val output: Map[String, Any] = Map("user_id"->uid, "business_id"->bid, "stars"->rating)
        val formatted_output = org.json4s.jackson.Serialization.write(output)
        out.write(formatted_output)
        out.write("\n")
      }
    }

    out.close()
  }

  def userBased(args: Array[String]) = {

  }

  def main(args: Array[String]) = {
    val t1 = System.nanoTime
    val cfType = args(4)

    if (cfType == "item_based") {
      itemBased(args)
    }
    else {
      userBased(args)
    }

    println("Duration: " + (System.nanoTime - t1) / 1e9d)
  }
}
